addons/web_editor/static/src/js/editor/image_processing.js

```diff
@@ -431,12 +431,18 @@ async function loadImageInfo(img, rpc, attachmentSrc = '') {
         return;
     }
 
+    // Check if the src is a local and absolute URL: if yes, only consider
+    // the "relative" part of the URL.
+    const locationOrigin = window.location.origin;
+    const isSrcAbsoluteUrl = src.startsWith(locationOrigin);
+    const relativeSrc = isSrcAbsoluteUrl ? src.split(locationOrigin)[1] : src;
```

Comment:
I would write this that way:
```suggestion
    const relativeSrc = src.split(window.location.host).pop();
```
- One line.
- Uses the `host` instead of the `origin`, that way this works if the protocol is omitted too.

Although in my opinion, we could do a bit more: it technically may have happened that users started in 16.0 without a domain name and use the grid mode while it was making the image URL go absolute then they added a domain later. This means that this would happen:
- Add an image with `/web/image/stuff` URL
- Use grid mode: it becomes `https://mycompany.odoo.com/web/image/stuff`
- Add the domain name: the website `mycompany.be` now uses `https://mycompany.odoo.com` image URLs, not pretty but still working
- Go through this new code => still not working you consider the image URL is not absolute here.

I don't think the check of `isLocal` is actually that useful below. Could you investigate on that? I mean:
1. First, it's already a weird case to put an absolute URL for an image that is outside of your domain (only way to do it should already be via the HTML editor? Although I think we had a bug recently where using the "Add URL" button in the media dialog added the URL directly in the DOM instead of creating an attachment for it... if I am not mistaken @rdeodoo).
2. If you just always get the relative URL (without checking if same origin), it's unlikely `get_image_info` will return anything (it would mean your outside-of-Odoo image would use the same pathname as one image in your database)
3. Even if that unlikely event happens... is something bad happening? We'll just consider that image to be the original one for this foreign image that the user could never have cropped/filter/whatever anyway since it is a foreign image? But we could be extra careful and only do that if the relative path starts with `/web/image` maybe and/or if the absolute url contained `odoo.com`. I let you investigate a bit what's best.

Reply:
> Although I think we had a bug recently where using the "Add URL" button in the media dialog added the URL directly in the DOM instead of creating an attachment for it... if I am not mistaken @rdeodoo).

A bit worst but yep, it was created the attachment but not using it at all, and indeed hardcoding the external image URL in the DOM.

Reply:
Note than I just had a look and:
- In 16.4 and + we merged [this](https://github.com/odoo/odoo/commit/943944dd249c15de870d6800d89e48d54a422e5a) which will download the image in the attachment and use it, so no more link
- Before that, we kept it as it is, I guess because we found this better way in master (16.4 at the time), so a bugfix was not that much needed for stable

Reply:
@qsm-odoo (see last force-push)
1) > One line.
Uses the host instead of the origin, that way this works if the protocol is omitted too. 

     =>  If the protocol is omitted, the url is considered as a pathname and is appended to the website url (ex: https://localhost:8069/www.odoo.com/logo) so I don't think it is a case to consider (I am saying that to justify what I say in point 2) 

2. > it technically may have happened that users started in 16.0 without a domain name and use the grid mode while it was making the image URL go absolute then they added a domain later
  
     => Indeed, so I made a solution that extracts the pathname from any URL :slightly_smiling_face: Note that I could have used a regex like this instead
 ```js
const domain = src.match(/^(?:https?:\/\/)?(?:[^@/\n]+@)?(?:www\.)?([^/\n]+)/g);
const relativeSrc = src.split(domain && domain[0]).pop();
```
  (I found it [here](https://regex101.com/r/jN6kU2/1) and modified it by removing the `:` in the last part so it considers the ports too)
  But I found the regex ugly and preferred to trust the URL API (considering what I said in point 1) :slightly_smiling_face:  

3. > I don't think the check of isLocal is actually that useful below. Could you investigate on that?
  
I tested the following cases:

  - A user sets the src as a complete absolute URL (complete = with protocol), either by adding an URL with the media dialog or by modifying the DOM with the html editor
    => since it is external, original is `false` so the `isLocal` check is useless 
    => **behavior**: nothing wrong happens, only the basic image options appears (so not the shapes, filter,...) The crop and transform options are here but there is a notification that appears when trying to crop, telling to upload the image before, and the transform works fine 
 
-  A user sets the src as a complete absolute URL with the html editor but he is modifying the Text-image image for example and forgets to delete the `data-original-*` attributes
 => since it has a `dataset.originalSrc`, `loadImageInfo` returns at the first if and so no fetch is performed (so `isLocal` is useless)
=> **behavior**: the image processing options (shape, filter) are here (because of the `data-original-*` attributes) but nothing wrong happens except that the image used for the transformations is the original one (so `originalSrc`)

-  A user sets the src as a complete absolute URL with a path that is the same as an existing image: if we consider only the relative path, then it is not that unlikely to have the same path as it seems that only the `id` is considered to find the original image (`web/image/<id>-<unique>/filename`) so it just needs to come from another `.../web/image/` with the same id to retrieve an original (if it exists)
 => `isLocal`  is useless as it returns true anyway
=> **behavior**: depending on how the image was added, it behaves differently but in all case, if the id is an existing one, an original is retrieved:
    - if we added it in the html editor, then the image is displayed and when we go in edit mode, the original is fetched and set as the `originalSrc` and is the one used by the image processing options
    - if we add the URL via the media dialog, then the original is retrieved and it replaces the "url image"
   
  => in all cases, the options work 

=> In conclusion, `isLocal` is useless when considering only the relative path directly.
However, maybe the relative path should only be considered when it is the current website (so as it was before) ? I could check for `.odoo.com` and `/web/image` like you said but then do you have an idea on how to distinguish the previous website domain (in the example you mentioned) and any other Odoo website ? Is the previous domain stored somewhere ? Or maybe it is very unlikely for an Odoo website to have an image of an other odoo website ? (I'll make the needed changes after your answer :slightly_smiling_face:)

Reply:
>  If the protocol is omitted, the url is considered as a pathname and is appended to the website url 

Yes, indeed :wink: But if you set up the URL with `//www.odoo.com/logo.png` instead `www.odoo.com/logo.png`, it will work and splitting on the `location.origin` would not have worked (or simply using a different protocol) :wink: In any case, using the URL api was the way to go :+1:

> A user sets the src as a complete absolute URL with the html editor but he is modifying the Text-image image for example and forgets to delete the data-original-* attributes

Good to think about it but if that was to be a problem we would have ignored it. If users "breaks the DOM structure" **, we should not support it :)

> it seems that only the id is considered to find the original image

I don't think it is the case, the code suggests that if we don't find one based on an ID, we try to match the whole URL based on the url field of ir.attachment records, no? `# Find attachment by url` in the code.

> maybe the relative path should only be considered when it is the current website (so as it was before) 

That's the question :wink: 

> how to distinguish the previous website domain (in the example you mentioned) and any other Odoo website ? Is the previous domain stored somewhere ?

No, of course, but the idea would have been that "Only if the image URL's domain contains odoo.com, we suppose we can treat them as a relative path". Like you said, very unlikely that odoo website use each other images for a legit reason. But I don't think it is needed, let's do as you made and see if someone ever complains. That would not be a critical bug: instead of having no possibility to crop (warning), no possibility to add a filter (no option?), etc... an image from their own database would be used as original if the external image had the same path as one image in DB. Unlikely and more "funny" than "buggy" if it were to happen IMO. But indeed, could happen more often than what I had in mind since we first match on ID only if it starts with `/web/image/` from what you are saying.

IMO, let's try like that. However, add a comment above your condition `if (original && original.image_src) {` to explain a bit of that (something like "If an absolute URL is used, we consider unlikely that the relative part matches something from this DB and even if it does, nothing bad would happen beside using a random image as original for options instead of having no option". Also explain why considering the relative part only is needed: because we converted relative URL by mistake at some point and user domains could have been changed meanwhile.

In the future (after OXP I guess, more urgent things to do, although if we do it for 17.0 I would be happy), I 100% want we do this anyway (if not impossible for some reason I am not seeing now):
- Send the whole URL to the `get_image_info` controller and let him check what's needed to be checked (he could actually compare the correct database odoo.com domain).
- Remove data-original-id and other things like that from the DOM, it's really a mess for no reason in the end:
    - More DOM to load for visitors
    - More things to handle for coders
    - More things to be wary about when using the HTML editor as an user as you explained **
    - ... all of this to avoid one RPC when clicking for the first time on an image that is being *re-edited* later. I think we just went crazy when we did that :open_mouth:

Reply:
> if we don't find one based on an ID, we try to match the whole URL based on the url field of ir.attachment records, no? # Find attachment by url in the code.

Ah indeed thanks for the info :slightly_smiling_face:  (What I really meant was that having `web/image/id....`  with the same id as an existing one was enough to have a match when I tested :slightly_smiling_face:  but you said that later so I think we agree haha )